src/modules/ods/ods.service.ts

```diff
@@ -0,0 +1,90 @@
+import { Injectable, InternalServerErrorException, NotFoundException } from '@nestjs/common';
+import { GetOdsCustomerResponse } from './dto/get-ods-customer-response.dto';
+import { InjectDataSource } from '@nestjs/typeorm';
+import { DATABASE } from '@ukef/constants';
+import { DataSource } from 'typeorm';
+import { PinoLogger } from 'nestjs-pino';
+import { ODS_ENTITIES, OdsEntity, odsStoredProcedureInput, odsStoredProcedureQueryParams } from './dto/ods-payloads.dto';
+
+@Injectable()
+export class OdsService {
+  constructor(
+    @InjectDataSource(DATABASE.ODS)
+    private readonly odsDataSource: DataSource,
+    private readonly logger: PinoLogger,
+  ) {}
+
+  async findCustomer(partyUrn: string): Promise<GetOdsCustomerResponse> {
+    try {
+      const spInput = this.createOdsStoredProcedureInput(ODS_ENTITIES.CUSTOMER, { customer_party_unique_reference_number: partyUrn });
+
+      const storedProcedureResult = await this.callOdsStoredProcedure(spInput);
+
+      const storedProcedureJson = JSON.parse(storedProcedureResult[0]?.output_body);
+
+      if (storedProcedureJson == undefined || storedProcedureJson?.status != 'SUCCESS') {
+        this.logger.error('Error from ODS stored procedure, output: %o', storedProcedureResult);
+        throw new InternalServerErrorException('Error trying to find a customer');
+      }
+
+      if (storedProcedureJson?.total_result_count == 0) {
+        throw new NotFoundException('No matching customer found');
+      }
+
+      return {
+        customerUrn: storedProcedureJson.results[0]?.customer_party_unique_reference_number,
+        customerName: storedProcedureJson.results[0]?.customer_name,
+      };
+    } catch (err) {
+      if (err instanceof NotFoundException) {
```

Comment:
the catching and rethrowing feels a little clunky here but is the best way to get better logging for debugging which I think is worth it. Any recommendations on refactoring this are welcome

Reply:
Its not ideal but to be honest, I can see reminisense of this pattern elsewhere in the codebase. Ideally an exceptional class will be clean but might creep the scope of this PR.